core/distributed/preconditioner/schwarz.cpp

```diff
@@ -84,15 +91,27 @@ void Schwarz<ValueType, LocalIndexType, GlobalIndexType>::apply_impl(
 {
     precision_dispatch_real_complex_distributed<ValueType>(
         [this](auto dense_alpha, auto dense_b, auto dense_beta, auto dense_x) {
-            auto x_clone = dense_x->clone();
-            this->apply_dense_impl(dense_b, x_clone.get());
+            set_cache_to(dense_x);
+            this->apply_impl(dense_b, cache_.intermediate.get());
             dense_x->scale(dense_beta);
-            dense_x->add_scaled(dense_alpha, x_clone.get());
+            dense_x->add_scaled(dense_alpha, cache_.intermediate.get());
         },
         alpha, b, beta, x);
 }
 
 
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+template <typename VectorType>
+void Schwarz<ValueType, LocalIndexType, GlobalIndexType>::set_cache_to(
```

Comment:
You only need this function once, right? Then I would prefer to just add this directly, it's not that much IMO.